Xena: overcloud DIB host image, LVM, swap & growroot

markgoddard opened the pull request.

markgoddard committed:
Add swap.yml custom playbook

Creates a swap filesystem on the lv_swap LV in the rootvg VG.

May be added as a post hook to overcloud host configure.

markgoddard committed:
Add growroot.yml custom playbook

Grows the root partition & rootpv LVM PV to fill the available space.

This playbook may be added as a pre-hook for overcloud host configure.

markgoddard committed:
Add StackHPC overcloud DIB configuration

markgoddard committed:
Add StackHPC LVM configuration

oneswig commented on etc/kayobe/inventory/group_vars/all/stackhpc-lvm:
```diff
@@ -0,0 +1,30 @@
+# StackHPC LVM lv_root LV size.
+stackhpc_lvm_lv_root_size: 50g
+
+# StackHPC LVM lv_tmp LV size.
```
Was there an issue with glance image conversions using /tmp?  We have some users who make very large images, potentially bigger than this.

markgoddard replied:
Wouldn't that be using /tmp inside the glance_api container? Which I think would use the container overlay filesystem by default.

markgoddard commented on etc/kayobe/stackhpc-lvm.yml:
```diff
@@ -0,0 +1,117 @@
+---
```
Might need to move all of this to group_vars, since it's not possible to override these variables on a per-group basis. e.g. `stackhpc_lvm_group_rootvg_disks`, and `stackhpc_lvm_group_rootvg_lvs_extra`.

priteau commented on etc/kayobe/stackhpc-overcloud-dib.yml:
```diff
@@ -0,0 +1,212 @@
+
+# StackHPC overcloud DIB image elements.
+stackhpc_overcloud_dib_elements:
+  - "{{ os_distribution }}-minimal"
```
This won't work for rocky since there isn't a rocky-minimal element.

markgoddard commented on etc/kayobe/ansible/growroot.yml:
```diff
@@ -0,0 +1,51 @@
+      become: True
+
+    - name: Get root PV device
+      command: "pvs --select vg_name=rootvg --reportformat json"
```
@priteau says please make it a variable

markgoddard commented on etc/kayobe/ansible/swap.yml:
```diff
@@ -0,0 +1,26 @@
+---
+- name: Configure swap
+  hosts: overcloud
```
Make optional

jovial replied:
Not all hosts may want swap

markgoddard commented on etc/kayobe/overcloud-dib.yml:
```diff
@@ -10,6 +10,10 @@
 
+# List of additional host packages to install.
+overcloud_dib_host_packages_extra:
+  - "{% if 'ubuntu-minimal' in overcloud_dib_host_images | map(attribute='elements') | flatten | list %}debootstrap{% endif %}"
```
This requires a backport of https://review.opendev.org/c/openstack/kayobe/+/850903 in Kayobe.

priteau replied:
This hasn't merged yet although the depending change has and Workflow+1 has been added. I am asking on IRC.

priteau replied:
This should be fixed by https://review.opendev.org/c/openstack/project-config/+/853003.

markgoddard replied:
https://github.com/stackhpc/kayobe/pull/39

markgoddard commented:
Addressed comments, fixed up commits, and rebased onto the `xena-wallaby-merge` branch.

If you would like to compare against the original branch, it is `xena-overcloud-dib-orig`.

All 3 distros build, and deploy in VMs on SMS lab with the expected partition layout. I didn't test bare metal because of the NIC drivers.

markgoddard commented on etc/kayobe/stackhpc-overcloud-dib.yml:
```diff
@@ -0,0 +1,212 @@
+# Example:
+# overcloud_dib_host_images:
+#   - "{{ stackhpc_overcloud_dib_host_image }}"
+stackhpc_overcloud_dib_host_image:
```
Wondering if this name should be any more specific than this? Will this be the only image we provide?

mnasiadka replied:
I guess let's start with that and see if we need more (hope not).
Do we plan to build those images in some CI and upload to Ark? Would help environments without ,,normal'' internet access.

markgoddard replied:
Yes. I have a workflow to build the images in progress, but not uploading to Ark yet. Probably @jackhodgkiss will look at that.

jovial commented on etc/kayobe/ansible/swap.yml:
```diff
@@ -0,0 +1,35 @@
+        fstype: "swap"
+        state: "present"
+
+    # It does no harm to run this when swap is already active
```
Why use the:
```
      when:
        - ansible_facts.swaptotal_mb == 0
```
condition then?

markgoddard replied:
Because that was already in the playbook when I picked it up :)

I guess it also helps with not showing things as changed. There may be cases where we avoid activating swap though, e.g. if facts are stale, or there is another swap device/file.

Should we remove it?

jovial replied:
Ahh, yeah, of course... For idempotency since that task will always report a changed status. Makes sense, thanks. I did find that in verbose mode swapon produces some output you could parse:
```
[will@juno ~]$ sudo swapon -av
swapon: /swapfile: found signature [pagesize=4096, signature=swap]
swapon: /swapfile: pagesize=4096, swapsize=2147483648, devsize=2147483648
swapon /swapfile
[will@juno ~]$ sudo swapon -av
swapon: /swapfile: already active -- ignored
```
to set changed_when, but feels like bike shedding.

jovial commented on etc/kayobe/ansible/growroot.yml:
```diff
@@ -0,0 +1,62 @@
+  tasks:
+    - name: Ensure growpart is installed
+      package:
+        name: "{% if os_distribution in ['centos', 'rocky'] %}cloud-utils-growpart{% else %}cloud-guest-utils{% endif %}"
```
This uses facts but the playbook doesn't collect them:

```
- name: Grow root PV
  hosts: "{{ growroot_group | default('overcloud') }}"
  # Avoid using facts because this may be used as a pre overcloud host
  # configure hook, and we don't want to populate the fact cache (if one is in
  # use) with the bootstrap user's context.
  gather_facts: false
```
Noticed this when trying to run it. Can we invalidate or refresh the cache for the kayobe playbooks?

jovial replied:
I seems you can use `meta: clear_facts` to clear the fact cache, but I guess clearing the cache is a nasty side effect!

jovial replied:
Possibly we can use `setup` with `gather_subset` to only collect the ones we need?

markgoddard replied:
Good catch. I'll use facts, with a fallback to `os_distribution`.

markgoddard committed:
growpart.yml: Don't assume facts are present

markgoddard committed:
Overcloud DIB: stop using StackHPC package repos for Ubuntu image

We are not currently syncing Ubuntu packages, since the on_demand mirror
in Ark does not work if the upstream mirror pulls packages (which it
does sometimes).

markgoddard committed:
growroot.yml: Fail if the expected volume group doesn't exist

markgoddard committed:
growroot: Avoid package installation when growpart is installed

In a pre-bootstrapped environment which requires a proxy for package
installation, we may not be able to perform package updates.
